### PR TITLE
fix: same TCP connection appears twice

### PR DESCRIPTION
**Pull request by weidongkl**
/proc/net/{TCP, TCP6, UDP, UDP6} are dynamic ,
and when we read these files, we should read them all at once.
there will be data consistency issues if using line by line reading

fix: https://github.com/prometheus/procfs/issues/576

### CODE REVIEW

**Comment by weidongkl** [comment]
net-tools used  setvbuf _IOFBF to read tcp file at once too. data consistency should be more important than the increased memory usage

**Comment by weidongkl** [comment]
> @SuperQ wdyt

**Comment by discordianfish** [comment]
Yeah I think we'll need to do this

**Review by discordianfish — APPROVED** [review]

File: net_ip_socket.go [hunk 1]
```diff
@@ -74,49 +62,50 @@ type (
...
-	f, err := os.Open(file)
-	if err != nil {
-		return nil, err
-	}
-	defer f.Close()
-
 	var netIPSocketSummary NetIPSocketSummary
 	var udpPacketDrops uint64
 	isUDP := strings.Contains(file, "udp")
+	content, err := os.ReadFile(file)
```
Comment (SuperQ) [inline net_ip_socket.go:95]: Would we want to use the `util.ReadFileNoStat()` function here?

Comment (SuperQ) [reply net_ip_socket.go:95]: Ahh, nevermind, this file is beyond the 1MiB file size limit of that util function.

**Comment by dswarbrick** [comment]
> net-tools used setvbuf _IOFBF to read tcp file at once too. data consistency should be more important than the increased memory usage

net-tools does not read the entire file at once. It merely uses buffered IO, with a buffer of PAGE_SIZE:
```c
FILE *proc_fopen(const char *name)
{
    static char *buffer;
    static size_t pagesz;
    FILE *fd = fopen(name, "r");

    if (fd == NULL)
      return NULL;

    if (!buffer) {
      pagesz = getpagesize();
      buffer = malloc(pagesz);
    }   

    setvbuf(fd, buffer, _IOFBF, pagesz);
    return fd; 
}
```
I suspect that an equivalent approach in Go, which would have only required a single line change, would be to replace the previous `io.LimitReader` with a `bufio.NewReaderSize(f, os.Getpagesize())`.

**Comment by SuperQ** [comment]
Actually, this may be worse. `os.ReadFile()` attempts to open and read into a buffer the size of the file in one read. But only if it can stat the size of the file. Otherwise it reads 512 bytes at a time.

The proc files in question have no size.
```
$ stat /proc/net/tcp
  File: /proc/net/tcp
  Size: 0         	Blocks: 0          IO Block: 1024   regular empty file
Device: 17h/23d	Inode: 4026532077  Links: 1
Access: (0444/-r--r--r--)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2024-04-16 09:48:28.233648680 +0200
Modify: 2024-04-16 09:48:28.233648680 +0200
Change: 2024-04-16 09:48:28.233648680 +0200
 Birth: -
```

So this change likely does not fix the problem and may make it worse.

I think I'm going to revert this, without proof that this fix actually works, I think the original code is actually what we want.

**Comment by dswarbrick** [comment]
Also for comparison, the (deprecated) `netstat` command:
```
openat(AT_FDCWD, "/proc/net/tcp", O_RDONLY) = 3
read(3, "  sl  local_address rem_address "..., 4096) = 4050
read(3, "  26: 0601A8C0:8938 8416AB51:05A"..., 4096) = 4050
read(3, "  53: 0601A8C0:8102 182F853E:589"..., 4096) = 4050
read(3, "  80: 0601A8C0:B44A B6126749:D98"..., 4096) = 4050
read(3, " 107: 0601A8C0:8890 6D39D766:DE9"..., 4096) = 4050
read(3, " 134: 0601A8C0:A3A4 9630C658:1AE"..., 4096) = 300
read(3, "", 4096)                       = 0
close(3)
```
So if anything, the original code most closely resembles how net-tools reads /proc/net/tcp.

Now, if _both_ prometheus/procfs _and_ net-tools produce inconsistent / duplicated entries, then I would say that it's time to use rtnetlink to get this information, as iproute2 does with the `ss` (socket statistics) tool.

**Comment by SuperQ** [comment]
We already switched to netlink in the node_exporter tcpstat collector.

**Comment by dswarbrick** [comment]
Exactly - so should these legacy functions that parse `/proc/net/tcp{,6}` be marked as deprecated to discourage people from using them?

**Comment by SuperQ** [comment]
Yea, maybe not a bad idea. @discordianfish What do you think of marking this reader as deprecated?

**Comment by discordianfish** [comment]
@SuperQ yeah that makes sense

**Pull request merged**